### PR TITLE
Add parent class support from reflection, remove properties, methods …

### PR DESCRIPTION
**Pull request by cotcotquedec**
…and traits management from extended class, add export methods for write array a php 5.4 syntax

### CODE REVIEW

File: src/model/PhpClass.php [hunk 1]
```diff
@@ -36,10 +36,14 @@ class PhpClass extends AbstractPhpStruct implements GenerateableInterface, Trait
 	 * @return PhpClass
 	 */
 	public static function fromReflection(\ReflectionClass $ref) {
+
+		/**@var PhpClass $class */
```
Comment (cotcotquedec) [inline src/model/PhpClass.php]: For IDE autocompletion

File: src/model/PhpClass.php [hunk 2]
```diff
@@ -36,10 +36,14 @@ class PhpClass extends AbstractPhpStruct implements GenerateableInterface, Trait
...
 		$class = new static();
+
 		$class->setQualifiedName($ref->name)
 			->setAbstract($ref->isAbstract())
 			->setFinal($ref->isFinal())
+			->setParentClassName($ref->getParentClass()->name)
```
Comment (cotcotquedec) [inline src/model/PhpClass.php]: Use Parent class!

File: src/model/PhpClass.php [hunk 3]
```diff
@@ -51,17 +55,23 @@ public static function fromReflection(\ReflectionClass $ref) {
 
 		// methods
 		foreach ($ref->getMethods() as $method) {
-			$class->setMethod(static::createMethod($method));
+			if ($method->class == $ref->name) {
```
Comment (cotcotquedec) [inline src/model/PhpClass.php]: Don't wanna generated class with method from parent class

File: src/model/PhpProperty.php [hunk 4]
```diff
@@ -42,6 +42,34 @@ public static function create($name) {
...
 
+
+	/**
+	 * Export array as php 5.4
+	 *
+	 * @param $var
+	 * @param string $indent
+	 * @return string
+	 */
+	static public function export($var, $indent="") {
```
Comment (cotcotquedec) [inline src/model/PhpProperty.php]: Array management for property

File: src/model/PhpProperty.php [hunk 5]
```diff
@@ -65,10 +93,14 @@ public static function fromReflection(\ReflectionProperty $ref) {
 		}
 
 		$defaultProperties = $ref->getDeclaringClass()->getDefaultProperties();
+
 		if (isset($defaultProperties[$ref->name])) {
 			$default = $defaultProperties[$ref->name];
 			if (is_string($default)) {
 				$property->setValue($default);
+			} elseif(is_array($default)) {
```
Comment (cotcotquedec) [inline src/model/PhpProperty.php]: Array management for property

File: src/model/PhpProperty.php [hunk 6]
```diff
@@ -65,10 +93,14 @@ public static function fromReflection(\ReflectionProperty $ref) {
...
+				$property->setValue(static::export($default));
+			} elseif(is_object($default)) {
```
Comment (cotcotquedec) [inline src/model/PhpProperty.php]: Il a property has no valu but has type, generation won't work, fix it

File: src/utils/Writer.php [hunk 7]
```diff
@@ -73,21 +73,25 @@ public function writeln($content = '') {
 	 * @param string $content        	
 	 */
 	public function write($content) {
-		$lines = explode("\n", $content);
-		for ($i = 0, $c = count($lines); $i < $c; $i++) {
-			if ($this->indentationLevel > 0
+
+		if (is_string($content)) {
+			$lines = explode("\n", $content);
```
Comment (cotcotquedec) [inline src/utils/Writer.php]: Array management, explode only on string

**Comment by gossi** [comment]
Thanks a lot. I really like the bugfixes for reading classes from reflection, especially with parent classes, as I didn't cover them. What about interfaces and traits? Will they also inherit methods that are available through reflection? Could you check this situation, too and provide tests for reflection with parents?

Second, the php5.4 array syntax: I'm a bit wondering how you managed to achieve what you did. Probably because you don't know the separation of concerns of this lib (as they are in my mind ;). The most "suspicious" method I'm wondering about is the static *export* on `PhpProperty`. Since values (and expressions are handled on the `ValuePart`) there is now `value` and `expression`.

Here are some rules of the lib:

- The model only consists of information
- Parsing/reading is either done in the parser or the `fromReflection` method but nowhere else
- Writing is done in the `GeneratorVisitor` and nowhere else

The `Writer` is "dumb", means it only flushes what it is been told. You teached him some logic, which isn't the right place here. This should be in the `GeneratorVisitor` and passing the right value to `Writer`.

Arrays: Arrays in php are so tricky in how they can look. It's also too risky to have an algorithm to parse/write them, as there is always a case a user wants that isn't catched in the algorithm. That's why there is now `setValue()` and `setExpression()`. The first is almost always escaped (through `var_export()`). However, if you want more control over your output, there is expression, which is never escaped, so you can pass code (especially useful for writing short php code statements and arrays as well, too).

So, could you rewrite your enhancements with this new knowledge?
Please feel free to ask questions.

**Comment by cotcotquedec** [comment]
If you want, you can find the very early alpha of my project here 
https://github.com/FrenchFrogs/maker

**Comment by gossi** [comment]
This can be closed as all the issues are being resolved with version 0.4.

**Pull request closed without merging**